File: 백준/Gold/27172. 수 나누기 게임/수 나누기 게임.py
import sys
input = lambda: sys.stdin.readline().rstrip()
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)

def main():
    n = int(input())
    card = [*map(int, input().split())]
    score = [0]*n
    maxv = max(card)
    loc = [-1]*(maxv+1)
    for i, v in enumerate(card): loc[v] = i
    for i, v in enumerate(card):
        for j in range(v*2, maxv+1, v):
            if loc[j] != -1:
                score[i] += 1
                score[loc[j]] -= 1
    print(*score)
    return
if __name__ == "__main__":
    main()

File: 백준/Gold/27172. 수 나누기 게임/수 나누기 게임.cc
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n, maxv = 0; cin >> n;
    vector<int> card(n), score(n, 0), loc(1'000'000, -1);
    for(int i = 0; i < n; i++){
        cin >> card[i];
        loc[card[i]] = i;
        if(maxv < card[i]) maxv = card[i];
    }
    for(int i = 0; i < n; i++){
        int cur = card[i];
        for(int j = cur*2; j <= maxv; j += cur) if(loc[j] != -1){
            score[i]++;
            score[loc[j]]--;
        }
    }
    for(int i = 0; i < n; i++) cout << score[i] << ' ';
    return 0;
}